test/logstream/v2/stream_test.go

```diff
@@ -145,15 +149,25 @@ func TestNamespaceStream(t *testing.T) {
 	// Kill the context.
 	cancel()
 
-	// Re-create pod, but the watch cycle must have finished by now.
-	if _, err := podClient.Create(context.Background(), pod, metav1.CreateOptions{}); err != nil {
-		t.Fatal("CreatePod()=", err)
-	}
-
-	select {
-	case <-time.After(time.Second):
-	case <-logFuncInvoked:
-		t.Error("No watching should have happened.")
+	// We can't assume that the cancel signal doesn't race the pod creation signal, so
+	// we retry a few times to give some leeway.
+	if err := wait.PollImmediate(10*time.Millisecond, time.Second, func() (bool, error) {
+		if _, err := podClient.Create(context.Background(), pod, metav1.CreateOptions{}); err != nil {
+			return false, err
+		}
+
+		select {
+		case <-time.After(noLogTimeout):
+			return true, nil
+		case <-logFuncInvoked:
+			t.Log("Log was still produced, trying again...")
+			if err := podClient.Delete(context.Background(), pod.Name, metav1.DeleteOptions{}); err != nil {
+				return false, err
+			}
+			return false, nil
+		}
+	}); err != nil {
+		t.Fatal("No watching should have happened", err)
 	}
```

Comment:
Dunno if this is a factor but the logFunc in the test is now blocking

```go
logFunc := func(format string, args ...interface{}) {
	logFuncInvoked <- struct{}{}
}
```

While it's blocked it blocks cancellation
https://github.com/knative/pkg/blob/eafc8c6f72a7aad7d1110e19f9d3d6cfbb7c0332/test/logstream/v2/stream.go#L198-L199

Reply:
I don't think that'd be a factor for the test honestly. The test's race is a race of us cancelling the context vs. creating the pod. If both signals arrive at the same time (from the PoV of the `select` statement), one will be picked "randomly".

Reply:
then there's a bug somewhere

`cancel()` is synchronous so `logFuncInvoked` should _never_ be invoked

Reply:
Cancel is, but is producing the signal onto the channel synchronous too?

Reply:
realized cancel was for the context and not the specific stream